solana/bridge_token_factory/programs/bridge_token_factory/src/instructions/user/register_mint.rs

```diff
@@ -0,0 +1,136 @@
+use anchor_lang::prelude::*;
+use anchor_spl::{
+    associated_token::AssociatedToken,
+    metadata::MetadataAccount as MplMetadata,
+    token_2022::{
+        self,
+        spl_token_2022::{
+            self,
+            extension::{
+                metadata_pointer::MetadataPointer, BaseStateWithExtensions, StateWithExtensions,
+            },
+        },
+    },
+    token_interface::{
+        spl_token_metadata_interface::state::TokenMetadata, Mint, TokenAccount, TokenInterface,
+    },
+};
+
+use crate::error::ErrorCode;
+use crate::instructions::wormhole_cpi::*;
+use crate::{
+    constants::{AUTHORITY_SEED, VAULT_SEED},
+    state::message::{deploy_token::DeployTokenPayload, Payload},
+};
+use anchor_spl::metadata::ID as MetaplexID;
+
+#[derive(Accounts)]
+pub struct RegisterMint<'info> {
+    #[account(
+        seeds = [AUTHORITY_SEED],
+        bump = wormhole.config.bumps.authority,
+    )]
+    pub authority: SystemAccount<'info>,
+
+    #[account(
+        constraint = !mint.mint_authority.contains(authority.key),
+        mint::token_program = token_program,
+    )]
+    pub mint: Box<InterfaceAccount<'info, Mint>>,
+
+    pub override_authority: Option<Signer<'info>>,
+
+    #[account()]
+    pub metadata: Option<Account<'info, MplMetadata>>,
+
+    #[account(
+        init,
+        payer = wormhole.payer,
+        token::mint = mint,
+        token::authority = authority,
+        seeds = [
+            VAULT_SEED,
+            mint.key().as_ref(),
+        ],
+        bump,
+        token::token_program = token_program,
+    )]
+    pub vault: Box<InterfaceAccount<'info, TokenAccount>>,
+
+    pub wormhole: WormholeCPI<'info>,
+
+    pub system_program: Program<'info, System>,
+    pub token_program: Interface<'info, TokenInterface>,
+    pub associated_token_program: Program<'info, AssociatedToken>,
+}
+
+#[derive(AnchorSerialize, AnchorDeserialize)]
+pub struct MetadataOverride {
+    pub name: String,
+    pub symbol: String,
+}
+
+impl<'info> RegisterMint<'info> {
+    pub fn process(&mut self, metadata_override: MetadataOverride) -> Result<()> {
+        let (name, symbol) = if let Some(override_authority) = self.override_authority.as_ref() {
+            match override_authority.key() {
+                a if a == self.wormhole.config.admin => {}
+                a if self.mint.mint_authority.contains(&a) => {}
+                _ => return err!(ErrorCode::Unauthorized),
+            }
+            (metadata_override.name, metadata_override.symbol)
```

Comment:
@kiseln 
Do I understand correctly that the metadata should be passed manually by the admin of the contract?
If yes, then this flow overcomplicate the deploy process on Solana, the name and symbol can be empty, and the metadata can be set by the DAO on the NEAR side.

Reply:
@aankor

Reply:
No, this is optional. Metadata is retrieved automatically but admin can override